tests/integration/test_headers_and_subtotals.py

```diff
@@ -374,6 +375,247 @@ def test_subtotals_proportions_2d_cube_with_hs_on_two_dim_by_cell(self):
+        expected = np.array([324, 198, 204, 121, 77, 49, 126])
+        np.testing.assert_almost_equal(slice_.rows_margin, expected)
+
+    def test_subtotals_proportions_2d_cube_with_diffs_and_totals_on_both_by_cell(self):
+        slice_ = Cube(CR.CAT_SUBTOT_SUBDIFF_X_CAT_SUBTOT_SUBDIFF).partitions[0]
+        expected = [
+            [
+                -0.1679507,
+                -0.0323575,
+                -0.0123267,
+                -0.0446841,
+                0.0200308,
+                0.0847458,
+                0.0508475,
+            ],
+            [
+                -0.0955316,
+                0.0631741,
+                0.0416025,
+                0.1047766,
+                0.0416025,
+                0.0970724,
+                0.0616333,
+            ],
+            [
+                -0.1032357,
+                0.0662558,
+                0.0539291,
+                0.1201849,
+                0.0246533,
+                0.0708783,
+                0.0986133,
+            ],
+            [0.0616333, 0.0832049, 0.0261941, 0.1093991, 0.05547, 0.009245, 0.0123267],
+            [
+                0.0539291,
+                0.0677966,
+                0.0323575,
+                0.1001541,
+                0.0046225,
+                0.0077042,
+                0.0061633,
+            ],
+            [0.01849, 0.027735, 0.0215716, 0.0493066, 0.0169492, 0.0046225, 0.0046225],
+            [
+                0.0724191,
+                0.0955316,
+                0.0539291,
+                0.1494607,
+                0.0215716,
+                0.0123267,
+                0.0107858,
+            ],
+        ]
+        np.testing.assert_almost_equal(slice_.table_proportions, expected)
+
+    def test_subtotals_proportions_2d_cube_with_diffs_and_totals_on_both_by_col(self):
+        slice_ = Cube(CR.CAT_SUBTOT_SUBDIFF_X_CAT_SUBTOT_SUBDIFF).partitions[0]
+        expected = [
+            [
+                -0.2466063,
+                -0.105,
+                -0.0701754,
+                -0.0923567,
+                0.1397849,
+                0.4471545,
+                0.2773109,
+            ],
+            [-0.1402715, 0.205, 0.2368421, 0.2165605, 0.2903226, 0.5121951, 0.3361345],
+            [-0.1515837, 0.215, 0.3070175, 0.2484076, 0.172043, 0.3739837, 0.5378151],
+            [0.0904977, 0.27, 0.1491228, 0.2261146, 0.3870968, 0.0487805, 0.0672269],
+            [0.0791855, 0.22, 0.1842105, 0.2070064, 0.0322581, 0.0406504, 0.0336134],
+            [0.0271493, 0.09, 0.122807, 0.1019108, 0.1182796, 0.0243902, 0.0252101],
+            [0.1063348, 0.31, 0.3070175, 0.3089172, 0.1505376, 0.0650407, 0.0588235],
+        ]
+        np.testing.assert_almost_equal(slice_.column_proportions, expected)
+
+    def test_subtotals_proportions_2d_cube_with_diffs_and_totals_on_both_by_row(self):
+        slice_ = Cube(CR.CAT_SUBTOT_SUBDIFF_X_CAT_SUBTOT_SUBDIFF).partitions[0]
+        expected = [
```

Comment:
These large expectations tend to render the code unreadable. Could you put them into a file? There are examples of it throughout the test suite.

Reply:
In general I think that a criterion should be "if it's more than a single screen hight - move it to a file". But I think @scanny has an even better rule, but I don't remember it now.